apps/examples/araplayer/HLSParser.cpp

```diff
@@ -0,0 +1,281 @@
+				segUrl[idx++] = *data++;
+			}
+
+			segUrl[idx] = 0;
+			playlist->vecUrl.push_back(std::string(segUrl));
+		} else {
+			++data;
+		}
+	}
+
+	playlist->parsingCnt++;
+	if (mCallbackObject != NULL && mCBGetPlaylistInfo != NULL) {
+		if (isFirstParsedMediaPlaylist(playlist)) {
+			pInfo.playlistUrl = playlist->parent->url;
+		} else {
+			pInfo.playlistUrl = url;
+		}
+
+		getMediaPlaylistInfo(playlist, pInfo);
+		mCBGetPlaylistInfo(mCallbackObject, pInfo.playlistUrl.c_str(), pInfo);
+	}
+
+	return 1;
+}
+
+int HLSParser::getPlaylistInfo(const char *url, int bandwidth, has_segment_info_t *outSegInfo)
+{
+	int ret = 0;
+	PlaylistIter iter = mMapMasterPlaylist.find(url);
+	if (iter == mMapMasterPlaylist.end() || iter->second->parsingCnt <= 0) {
+		DEBUG_TRACE(LOG_ERR, "there is no appropriate playlist with url: %s", url);
+		return -1;
+	}
+
+	std::map<std::string, hls_play_list_t *> &mapMediaPlaylist = iter->second->mapMediaPlaylist;
+	for (PlaylistIter iter = mapMediaPlaylist.begin(); iter != mapMediaPlaylist.end(); ++iter) {
+		hls_play_list_t *playlist = iter->second;
+		if (playlist->bandwidth <= bandwidth) {
+			outSegInfo->url = std::string(playlist->url);
+			outSegInfo->bandwidth = bandwidth;
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+int HLSParser::getSegmentInfo(const char *url, int idx, int bandwidth, has_segment_info_t *outSegInfo)
+{
+	PlaylistIter iter = mMapMasterPlaylist.find(url);
+	hls_play_list_t *playlist;
+
+	if (iter == mMapMasterPlaylist.end()) {
+		DEBUG_TRACE(LOG_ERR, "there is no appropriate playlist with url: %s", url);
+		return -1;
+	}
+
+	std::map<std::string, hls_play_list_t *> &mapMediaPlaylist = iter->second->mapMediaPlaylist;
+	if (mapMediaPlaylist.size() > 0) {
+		for (PlaylistIter iter = mapMediaPlaylist.begin(); iter != mapMediaPlaylist.end(); ++iter) {
+			playlist = iter->second;
+			if (playlist->bandwidth == bandwidth) {
+				if (idx >= (int)playlist->vecUrl.size()) {
+					DEBUG_TRACE(LOG_ERR, "index is larger than media playlist vector size, idx: %d, bandwidth: %d", idx, bandwidth);
+					return -1;
+				}
+				outSegInfo->url = std::string(playlist->vecUrl.at(idx));
+				outSegInfo->strRange = "";
+				outSegInfo->bandwidth = bandwidth;
+				return 1;
+			}
+		}
+	} else {
+		playlist = iter->second;
+		if (idx >= (int)playlist->vecUrl.size()) {
+			DEBUG_TRACE(LOG_ERR, "index is larger than media playlist vector size, idx: %d, bandwidth: %d", idx, bandwidth);
+			return -1;
+		}
+
+		outSegInfo->url = std::string(playlist->vecUrl.at(idx));
+		outSegInfo->strRange = "";
+		outSegInfo->bandwidth = bandwidth;
+		return 1;
+	}
+
+	DEBUG_TRACE(LOG_DEBUG, "there is no appropriate media playlist, idx: %d, bandwidth: %d, url: %s", idx, bandwidth, url);
+	return 0;
+}
+
+int HLSParser::getMediaPlaylistInfo(hls_play_list_t *playlist, play_list_info_t &pInfo)
+{
+	std::map<std::string, hls_play_list_t *> mapMediaPlaylist;
+	std::vector<int> &vecQuality = pInfo.vecQuality;
```

Comment:
It is not vector quality, related to bandwidth. naming is needed.

Reply:
@tool3210 could you explain?